### ISSUE TITLE
_app.getInitialProps has appContext.ctx.req undefined when using Link in production for pages that doesn't have getServerSideProps

### ISSUE THREAD

**Issue by pierluigi-giancola-blink** [vercel/next.js#32531]
### What version of Next.js are you using?

11.1.0

### What version of Node.js are you using?

16.1.0

### What browser are you using?

Chrome

### What operating system are you using?

Ubuntu 20.04

### How are you deploying your application?

next start

### Describe the Bug

When using `_app.getInitialProps` I need to fetch some data based on the host name (IP or domain name).

For doing that I access `appContext.ctx.req?.headers.host`.

`appContext.ctx.req?.headers.host` is undefined when I navigate to a page that doesn't export getServerSideProps, using Link in production (probably related to prefetching).

### Expected Behavior

I expect `appContext.ctx.req?.headers.host` to don't be undefined when navigating between pages

### To Reproduce

I made a repro while investigating in the problem.

https://github.com/pierluigi-giancola-blink/nextjs-initialprops-error

The first paragraph of the README should be enough to walk you through the bug and the workaround.

To sum up very quickly:
- there are 3 pages: page1, page2, page3
- page1: declare getServerSideProps and do some data fetching in it
- page2: doesn't declare getServerSideProps
- page3: declare getServerSideProps that do nothing

page1 and page3 works, page2 breaks and breaks the other pages too when navigate back from browser, the only way to make it work is refresh the page.

**Comment by balazsorban44** [comment]
When navigating client-side, there is no "request". You only get the request object when accessing a page directly.

See the docs: https://nextjs.org/docs/api-reference/data-fetching/getInitialProps#context-object

**Comment by pierluigi-giancola-blink** [comment]
thanks for your reply, I still don't get why I don't have `request` when having getInitialProps in _app. Shouldn't all pages become SSR?

**Comment by balazsorban44** [comment]
> For the initial page load, getInitialProps will run on the server only. getInitialProps will then run on the client when navigating to a different route via the next/link component or by using next/router. However, if getInitialProps is used in a custom _app.js, and the page being navigated to implements getServerSideProps, then getInitialProps will run on the server.

**Comment by pierluigi-giancola-blink** [comment]
Got it, thank you again :smile:

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
